### PR TITLE
Add shared_memory_size, max_swap and swappiness to @batch decorator

### PR DESCRIPTION
**Pull request by savingoyal**
Adds `sharedMemorySize`, `maxSwap` and `swappiness` parameters to batch jobs on ec2 platform.

### CODE REVIEW

**Comment by savingoyal** [comment]
@jasonge27 Can you verify that metaflow-R plays well with this change?

**Comment by savingoyal** [comment]
@russellbrooks Can you try this PR with your PyTorch data loader code? You can set the `shared_memory_size` in code via `@batch(shared_memory_size=foo)` or on the CLI via `python flow.py run --with batch:shared_memory_size=foo`.

**Comment by russellbrooks** [comment]
hmm it looks like the `linuxParameters` aren't getting added into the batch job request.

here's what I'm seeing:
* grabbed branch code: `pip install -e git+https://github.com/Netflix/metaflow.git@sharedmem#egg=metaflow`
* whipped up a quick test flow:
```
from metaflow import FlowSpec, batch, step


class SHMTest(FlowSpec):
    @batch(cpu=8, memory=000, shared_memory_size=10000, queue="memory")
    @step
    def start(self):
        import os

        print(os.popen("df -h").read())
        print(os.popen("ipcs -lm").read())

        self.next(self.end)

    @step
    def end(self):
        pass


if __name__ == "__main__":
    SHMTest()
```

Output - `/dev/shm` still 64MB shared memory:
image


Inspecting the job request itself with:
`aws batch describe-jobs --jobs 268064a0-481c-4326-bda5-545f8e348777`

Obfuscated acct #s, etc:
```
{
    "jobs": [
        {
            "jobArn": "arn:aws:batch:us-east-1:############:job/268064a0-481c-4326-bda5-545f8e348777",
            "jobName": "russell-SHMTest-30198-start-308589-0",
            "jobId": "268064a0-481c-4326-bda5-545f8e348777",
            "jobQueue": "arn:aws:batch:us-east-1:############:job-queue/memory",
            "status": "STARTING",
            "attempts": [],
            "createdAt": 1610059577337,
            "dependsOn": [],
            "jobDefinition": "arn:aws:batch:us-east-1:############:job-definition/metaflow_7b6b5ac7d4780531ec2d40674e46aa0224755690b9df387d8b01f546:1",
            "parameters": {
                "metaflow.step_name": "start",
                "metaflow.flow_name": "SHMTest",
                "metaflow.run_id": "30198",
                "metaflow.retry_count": "0",
                "metaflow.version": "2.2.5.post12+git23a99e3",
                "metaflow.task_id": "308589",
                "metaflow.user": "russell"
            },
            "container": {
                "image": "############.dkr.ecr.us-east-1.amazonaws.com/############",
                "command": [
                    "/bin/sh",
                    "-c",
                    "set -e && echo 'Setting up task environment.' && python -m pip install awscli click requests boto3 -qqq && mkdir metaflow && cd metaflow && mkdir .metaflow && i=0; while [ $i -le 5 ]; do echo 'Downloading code package.'; python -m awscli s3 cp s3://############/SHMTest/data/8d/8d3a41c020a30d99cc6ba89cc69868d84616ab43 job.tar >/dev/null &&                         echo 'Code package downloaded.' && break; sleep 10; i=$((i+1)); done && if [ $i -gt 5 ]; then echo 'Failed to download code package from s3://############/SHMTest/data/8d/8d3a41c020a30d99cc6ba89cc69868d84616ab43 after 6 tries. Exiting...' && exit 1; fi && tar xf job.tar && echo 'Task is starting.' && python -u shm_test_flow.py --quiet --metadata service --environment local --datastore s3 --event-logger nullSidecarLogger --monitor nullSidecarMonitor --datastore-root s3://############ --package-suffixes .py,.R,.RDS --pylint step start --run-id 30198 --task-id 308589 --input-paths ${METAFLOW_INPUT_PATHS_0}"
                ],
                "jobRoleArn": "arn:aws:iam::############:role/############",
                "volumes": [],
                "environment": [
                    {
                        "name": "METAFLOW_INPUT_PATHS_0",
                        "value": "30198/_parameters/308588"
                    },
                    {
                        "name": "METAFLOW_DATATOOLS_S3ROOT",
                        "value": "s3://############"
                    },
                    {
                        "name": "METAFLOW_SERVICE_URL",
                        "value": "https://06sqgps7tj.execute-api.us-east-1.amazonaws.com/api"
                    },
                    {
                        "name": "METAFLOW_CODE_DS",
                        "value": "s3"
                    },
                    {
                        "name": "METAFLOW_CODE_SHA",
                        "value": "8d3a41c020a30d99cc6ba89cc69868d84616ab43"
                    },
                    {
                        "name": "METAFLOW_DEFAULT_METADATA",
                        "value": "service"
                    },
                    {
                        "name": "METAFLOW_DATASTORE_SYSROOT_S3",
                        "value": "s3://############"
                    },
                    {
                        "name": "AWS_DEFAULT_REGION",
                        "value": "us-east-1"
                    },
                    {
                        "name": "METAFLOW_USER",
                        "value": "russell"
                    },
                    {
                        "name": "METAFLOW_SERVICE_HEADERS",
                        "value": "{\"x-api-key\": \"############\"}"
                    },
                    {
                        "name": "METAFLOW_DEFAULT_DATASTORE",
                        "value": "s3"
                    },
                    {
                        "name": "METAFLOW_CODE_URL",
                        "value": "s3://############/SHMTest/data/8d/8d3a41c020a30d99cc6ba89cc69868d84616ab43"
                    }
                ],
                "mountPoints": [],
                "ulimits": [],
                "containerInstanceArn": "arn:aws:ecs:us-east-1:############:container-instance/memory_Batch_33d9c6f1-e856-3b80-81f4-cf2e6cc79b6f/f04c37ad313948bb9d77f7c59cfa6259",
                "taskArn": "arn:aws:ecs:us-east-1:############:task/memory_Batch_33d9c6f1-e856-3b80-81f4-cf2e6cc79b6f/0041b9b531f548dba97362fd0c7f3d8c",
                "logStreamName": "metaflow_7b6b5ac7d4780531ec2d40674e46aa0224755690b9df387d8b01f546/default/0041b9b531f548dba97362fd0c7f3d8c",
                "networkInterfaces": [],
                "resourceRequirements": [
                    {
                        "value": "8",
                        "type": "VCPU"
                    },
                    {
                        "value": "60000",
                        "type": "MEMORY"
                    }
                ],
                "secrets": []
            },
            "timeout": {
                "attemptDurationSeconds": 432000
            },
            "tags": {},
            "platformCapabilities": []
        }
    ]
}
```

**Comment by savingoyal** [comment]
Can you check your `job definition`? `linuxParameters` are a job definition concept for AWS Batch.

**Comment by russellbrooks** [comment]
`aws batch describe-job-definitions --job-definition-name metaflow_7b6b5ac7d4780531ec2d40674e46aa0224755690b9df387d8b01f546`
```
{
    "jobDefinitions": [
        {
            "jobDefinitionName": "metaflow_7b6b5ac7d4780531ec2d40674e46aa0224755690b9df387d8b01f546",
            "jobDefinitionArn": "arn:aws:batch:us-east-1:############:job-definition/metaflow_7b6b5ac7d4780531ec2d40674e46aa0224755690b9df387d8b01f546:1",
            "revision": 1,
            "status": "ACTIVE",
            "type": "container",
            "parameters": {},
            "containerProperties": {
                "image": "############.dkr.ecr.us-east-1.amazonaws.com/############",
                "command": [
                    "echo",
                    "hello world"
                ],
                "jobRoleArn": "arn:aws:iam::############:role/############",
                "volumes": [],
                "environment": [],
                "mountPoints": [],
                "ulimits": [],
                "resourceRequirements": [
                    {
                        "value": "1",
                        "type": "VCPU"
                    },
                    {
                        "value": "4096",
                        "type": "MEMORY"
                    }
                ],
                "secrets": []
            },
            "tags": {}
        }
    ]
}
```

**Comment by savingoyal** [comment]
Screenshot 2021-01-07 at 3 14 30 PM

I do see /dev/shm being set correctly on my end.

**Comment by savingoyal** [comment]
And at the same time, I **don't** see `linuxParameters` populated for the job definition through `aws-cli`; however the UI correctly reflects the shared memory.

**Comment by savingoyal** [comment]
Aha, figured it out, can you make sure you are using the latest version of `awscli`? The older version is omitting entries in `linuxParameters`.

**Comment by savingoyal** [comment]
You might also have to update your version of `boto3` to the latest. That could be the reason why you are seeing the current behavior. Before you upgrade your version of `boto3`, can you let me know what version are you on?

**Comment by russellbrooks** [comment]
Excellent find!

FWIW before the UI for the job definition was empty as well:
image


I had `boto3    1.16.39` with:
```
$ aws --version
aws-cli/1.18.199 Python/3.7.7 Linux/5.4.0-1034-aws botocore/1.19.39
```

Upgrading awscli, leaving boto3:
```
$ pip install --upgrade awscli
$ aws --version
aws-cli/1.18.211 Python/3.7.7 Linux/5.4.0-1034-aws botocore/1.19.51
```

Testing with this still didn't work after upgrading boto3 to 1.16.51 as well.

Are you using awscli v2? Also is there something I should I modify to trigger a new job def version to be created?

**Comment by savingoyal** [comment]
Can you delete the previous job definition and then try?

**Comment by russellbrooks** [comment]
Deregistering the job definition and rerunning triggered a new revision to be created, but still not included 🙁

**Comment by savingoyal** [comment]
I am on the same version of `aws-cli` - `aws-cli/1.18.211 Python/3.7.7 Darwin/19.6.0 botocore/1.19.51` and I have `boto3==1.16.51` and `botocore==1.19.51`.

**Comment by russellbrooks** [comment]
Confirmed on the versions and the same interpreter, still no luck.

Something interesting when comparing the newly created job def revisions across the new and old batch dashboards too:
image

In the new dashboard there's nothing about the new batch multinode functionality:
image

 Compared to a previously created metaflow job definition:
image

**Comment by russellbrooks** [comment]
Ahhh I think I see it – looks to be this logic in the batch_client:
```
platform = response["computeEnvironments"][0]["computeResources"]["type"]
```
When I look at my compute environment, that is `SPOT` instead of the expected `FARGATE` or `EC2`.
image

**Comment by savingoyal** [comment]
Good catch! I will make the change.

**Comment by russellbrooks** [comment]
Cloned the branch locally to test it out. Worked without needing to deregister as expected (now I see that hash going on 😉 )! 
image
🎉

**Comment by savingoyal** [comment]
Great! I have updated the PR.

**Review by russellbrooks — APPROVED** [review]
👍  looks good to me – appreciate the quick turnaround!

File: metaflow/plugins/aws/batch/batch.py [hunk 1]
```diff
@@ -113,10 +113,14 @@ def create_job(
         image,
         queue,
         iam_role=None,
+        execution_role=None,
         cpu=None,
         gpu=None,
         memory=None,
         run_time_limit=None,
+        shared_memory_size=None,
```
Comment (tuulos) [inline metaflow/plugins/aws/batch/batch.py:121]: nit: you could rename `shared_memory_size` to just `shared_memory`, so it is consistent with `memory` and other attributes

Comment (savingoyal) [reply metaflow/plugins/aws/batch/batch.py]: Done!

**Pull request merged**